app/domain/authentication/authn_azure/xms_mirid.rb

```diff
@@ -0,0 +1,65 @@
+module Authentication
+  module AuthnAzure
+
+    Err = Errors::Authentication::AuthnAzure
+
+    class XmsMirid
+
+      def initialize(xms_mirid_token_field)
+        @xms_mirid_token_field = xms_mirid_token_field
+
+        @xms_mirid_hash = xms_mirid_hash
+        validate
+      end
+
+      def subscriptions
+        @xms_mirid_hash["subscriptions"]
+      end
+
+      def resource_groups
+        @xms_mirid_hash["resourcegroups"]
+      end
+
+      def providers
+        @xms_mirid_hash["providers"]
+      end
+
+      private
+
+      # we expect the xms_mirid claim to be in the format of /subscriptions/<subscription-id>/...
+      # therefore, we ignore the first slash of the xms_mirid claim and group the entries in key-value pairs
+      # according to fields we need to retrieve from the claim.
+      # ultimately, transforming "/key1/value1/key2/value2" to {"key1" => "value1", "key2" => "value2"}
+      def xms_mirid_hash
+        split_xms_mirid = @xms_mirid_token_field.split('/')
+
+        # accept also an xms_mirid field that doesn't start with a slash
+        if split_xms_mirid.first == ''
+          split_xms_mirid = split_xms_mirid.drop(1)
+        end
+
+        split_xms_mirid
+          .slice_before(/subscriptions|resourcegroups|providers/)
+          .each_with_object({}) do |part, hash|
+          key, val  = part.first, part.drop(1)
+          val       = val.first if val.length == 1
```

Comment:
I'm about to make a subtle point I didn't bother with earlier, but now that you've done the extraction of this object you might enjoy.  Notice the:
```ruby
val       = val.first if val.length == 1
```
violates the "avoid mutation" heuristic I made above.  Tbf, in this case I don't really mind it, but I point it out as something that will disappear automatically when we make another change...

Notice there is something funny about what we're doing here, in that we return a _hash with a non-uniform data type_.  You're allowed to do this in loosy-goosy ruby, but you shouldn't.  That is, the _natural_ data type being returned here is `Hash[string]: [string]` -- a hash whose values are arrays of strings.  That's what we get if we just delete line 45.  So we've done something funny by making it into "a hash whose values are sometimes strings and sometimes arrays of strings".

Now that we have a nice little value object, we don't need to commit this minor sin anymore.  Instead, we can move the knowledge that we want "subscriptions" and "resourcegroups" to be simple strings into their accessor methods.  And memoization allows us to do this without no performance hit either:

```ruby
      def subscriptions
        @subscriptions ||= @xms_mirid_hash["subscriptions"].first
      end

      def resource_groups
        @resource_groups ||= @xms_mirid_hash["resource_groups"].first
      end
```

Reply:
🤤indeed!